### PR TITLE
Extend PersistenceExtensions with "between" methods

### LINKED ISSUE #1333: [persistence] averageSince shouldn't return null
The JavaDoc says that the Items state will be returned if it has not changed since the DateTime specified, [but it returns null](https://github.com/openhab/openhab-core/blob/master/bundles/org.openhab.core.model.persistence/src/org/openhab/core/model/persistence/extensions/PersistenceExtensions.java#L468). This should be changed. Changing the docs would be easy, but I think this would be better functionality.

### LINKED ISSUE #2280: [persistence] Extend functions to work on a past interval
averageSince() etc work on an interval ranging from now back to a date that's specified as a parameter.
On a number of applications it would be useful to have it work on a time period from date A until date B with B != now.
Helpful for e.g. any sort of machine learning of estimates from past values.

I thought this should be relatively easy to implement as some sort of offset into the persistence data "blob" of an item, but I may be mistaken as I'm not a core programmer.

### PR DESCRIPTION
**Pull request by J-N-K**
Closes #1333 
Closes #2280 

Signed-off-by: Jan N. Klug <github@klug.nrw>

### CODE REVIEW

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java [hunk 1]
```diff
@@ -544,27 +813,65 @@ private static DecimalType internalAverageSince(Item item, Iterator<HistoricItem
...
-        Iterable<HistoricItem> result = getAllStatesSince(item, timestamp, serviceId);
+        return internalSum(item, timestamp, null, serviceId);
+    }
+
+    /**
+     * Gets the sum of the state of a given <code>item</code> between two certain points in time.
+     * The {@link PersistenceService} identified by the <code>serviceId</code> is used.
+     *
+     * @param item the item for which we will sum its persisted state values between <code>begin</code> and
+     *            <code>ewd</code>
```
Comment (cweitkamp) [inline bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:860]:
```suggestion
     *            <code>end</code>
```

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java [hunk 2]
```diff
@@ -544,27 +813,65 @@ private static DecimalType internalAverageSince(Item item, Iterator<HistoricItem
...
     public static DecimalType sumSince(Item item, ZonedDateTime timestamp) {
         return sumSince(item, timestamp, getDefaultServiceId());
     }
 
+    /**
+     * Gets the sum of the state of a given <code>item</code> between two certain points in time.
+     * The default persistence service is used.
+     *
+     * @param item the item for which we will sum its persisted state values between <code>begin</code> and
+     *            <code>ewd</code>
```
Comment (cweitkamp) [inline bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:829]:
```suggestion
     *            <code>end</code>
```

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java [hunk 3]
```diff
@@ -246,10 +362,25 @@ public static Boolean updatedSince(Item item, ZonedDateTime timestamp, String se
...
 
+    /**
+     * Gets the historic item with the maximum value of the state of a given <code>item</code> since
+     * a certain point in time. The default persistence service is used.
+     *
+     * @param item the item to get the maximum state value for
+     * @param begin the point in time to start the check
+     * @param end the point in time to stop the check
+     * @return a {@link HistoricItem} with the maximum state value since the given point in time, or
+     *         <code>null</code>aif no states found or if the default persistence service does not refer to an available
```
Comment (cweitkamp) [inline bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:377]:
```suggestion
     *         <code>null</code> if no states found or if the default persistence service does not refer to an available
```

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java [hunk 4]
```diff
@@ -212,12 +299,28 @@ public static Boolean changedSince(Item item, ZonedDateTime timestamp, String se
...
     }
 
     /**
-     * Checks if the state of a given <code>item</code> has changed since a certain point in time.
+     * Checks if the state of a given <code>item</code> has been updated between two points in time.
+     * The default persistence service is used.
+     *
+     * @param item the item to check for state updates
+     *
+     *
```
Comment (cweitkamp) [inline bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:312]:
```suggestion
     * @param begin the point in time to start the check
     * @param end the point in time to stop the check
```

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java [hunk 5]
```diff
@@ -262,11 +393,35 @@ public static HistoricItem maximumSince(Item item, ZonedDateTime timestamp) {
...
+     * a certain point in time. The {@link PersistenceService} identified by the <code>serviceId</code> is used.
+     *
+     * @param item the item to get the maximum state value for
+     * @param begin the point in time to start the check
+     * @param end the point in time to stop the check
+     * @param serviceId the name of the {@link PersistenceService} to use
+     * @return a {@link HistoricItem} with the maximum state value since the given point in time, or
+     *         <code>null</code>aif no states found or if the given <code>serviceId</code> does not refer to an
+     *         available
+     *         {@link QueryablePersistenceService}
```
Comment (cweitkamp) [inline bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:411]:
```suggestion
     *         <code>null</code> if no states found or if the given <code>serviceId</code> does not refer to an
     *         available {@link QueryablePersistenceService}
```

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java [hunk 6]
```diff
@@ -310,10 +445,25 @@ public String getName() {
...
     }
 
+    /**
+     * Gets the historic item with the minimum value of the state of a given <code>item</code> between
+     * two certain points in time. The default persistence service is used.
+     *
+     * @param item the item to get the minimum state value for
+     * @param begin the beginning point in time
+     * @param end the end point in time to
+     * @return the historic item with the minimum state value between the given points in time, or a<code>null</code> if
```
Comment (cweitkamp) [inline bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:459]:
```suggestion
     * @return the historic item with the minimum state value between the given points in time, or <code>null</code> if
```

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java [hunk 7]
```diff
@@ -323,13 +473,35 @@ public static HistoricItem minimumSince(Item item, ZonedDateTime timestamp) {
...
+
+    /**
+     * Gets the historic item with the minimum value of the state of a given <code>item</code> between
+     * two certain points in time. The {@link PersistenceService} identified by the <code>serviceId</code> is used.
+     *
+     * @param item the item to get the minimum state value for
+     * @param begin the beginning point in time
+     * @param end the end point in time to
+     * @param serviceId the name of the {@link PersistenceService} to use
+     * @return the historic item with the minimum state value between the given points in time, or a<code>null</code> if
```
Comment (cweitkamp) [inline bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:490]:
```suggestion
     * @return the historic item with the minimum state value between the given points in time, or <code>null</code> if
```

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java [hunk 8]
```diff
@@ -647,23 +896,39 @@ public static ZonedDateTime lastUpdate(Item item, String serviceId) {
...
+     * Gets the difference value of the state of a given <code>item</code> between two certain point in time.
+     * The default persistence service is used.
+     *
+     * @param item the item to get the delta for
+     * @param begin the beginning point in time
+     * @param end the end point in time
+     * @return the difference between end and begin, or <code>null</code> if the default persistence service does not
+     *         refer to an available {@link QueryablePersistenceService}, or if there is no persisted state for the
+     *         given
+     *         <code>item</code> for the given points in time
```
Comment (cweitkamp) [inline bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:913]:
```suggestion
     *         given <code>item</code> for the given points in time
```

**Review by cweitkamp — COMMENTED** [review]
I pointed out some spelling mistakes. Maybe you can check the JavaDoc formatting in the whole file. Thanks.

What about a `changedBetween(...)` method? Doesn't it work or is it to complex to be implemented?

**Comment by J-N-K** [comment]
Fixed JavaDoc and added `changedBetween`. Just didn't notice it's missing.

**Review by cweitkamp — APPROVED** [review]
LGTM. Thank you.

**Comment by J-N-K** [comment]
https://github.com/openhab/openhab-docs/pull/1835

**Comment by cweitkamp** [comment]
It looks like all these methods can be called with `start` > `end` value (see https://github.com/openhab/openhab-webui/pull/1634#issuecomment-1397744206). Should we add a check for that?

**Comment by J-N-K** [comment]
Yes, the question is: should we throw an exception or fix it by exchanging values?

**Comment by cweitkamp** [comment]
In my opinion we should log a warning and return `null` (or whatever the value is to be returned by the specific method). Similar to the handling if we are facing not a `QueryalblePersistenceService`.

**Pull request merged**